tensorflow_datasets/scripts/metadata_cleanup.py

```diff
@@ -0,0 +1,97 @@
+# Lint as: python3
+"""Removes all dataset versions from metadata directory which are not 
+   present in registered versions of TFDS.
+
+To test:
+Goto tensorflow_datasets/scripts
+
+Args : dry_run : If given, it traverse and finds all
+                 versions which are to be removed without
+                 actually removing them.
+
+             r : If given it removes all unmatched dataset versions 
+                 from metadata directory.
+```
+cd tensorflow_datasets/scripts
+
+For check not delete : python metadata_cleanup.py --dry_run
+For delete not check : python metadata_cleaup.py --r
+For both delete & check : python metadata_cleanup.py --dry_run --r
+```
+"""
+from absl import flags
+from absl import app
+import os
+import tensorflow as tf
+from tensorflow_datasets.core import registered
+
+
+FLAGS = flags.FLAGS
+flags.DEFINE_boolean('dry_run', False, "Shows all versions\
+                      which are not registered")
+flags.DEFINE_boolean('r', False, "Removes unmatched\
+                      dataset versions from metadata")
+
+__metadata_path = os.path.join(os.pardir, "testing/metadata")
+
+def _extract_metadata_versions(metadata_dir):
+    '''
+       Get all metadata direcotry versions paths,
+       It only extract the paths like 'dataset_name/version'
+       or 'dataset_name/config/versions' in metadata dir
+    '''
+    _meta_paths = set()
+    for root, dirs, files in tf.io.gfile.walk(metadata_dir):
+        for fileName in files:
+            _meta_paths.add(os.path.join(root[len(metadata_dir)+1:]))

```

Comment:
I see what you've done but this would only work if metadata do not contains sub-directory. Otherwise it might accidentally remove metadata from other folders.

It would be safer to check the path format.

Reply:
No its work fine I tested it like in Wikipedia directory `testing\metadata\wikipedia\20190301.hz\0.0.3`  it removes all files as this is not present in register versions

Reply:
I meant that ` testing\metadata\wikipedia\20190301.hz\0.0.3\som_dir` would be deleted too even if they should not. Ideally we should add some `_validate_name(root[len(metadata_dir)+1:])` which would check the name is of format `x/y/a.b.c`.

Reply:
Thanks for explanation. Yes it removes it too but is there is any `testing\metadata\wikipedia\20190301.hz\0.0.3\som_dir` directory is present inside versions like `0.0.3` in metadata ?
If so I have to match patterns as you suggest and as I was done in my previous commit?

Reply:
We want to make sure we won't create bug in the future.

To match the pattern, I would add a new `is_full_name` function to register, and would make sure to add unittests to check this works properly.

```
register.is_full_name('ds/config/1.0.2/other') -> False
register.is_full_name('ds/config/1.0.2/') -> False
register.is_full_name('ds/config/1.2') -> False
register.is_full_name('ds/config') -> False
register.is_full_name('ds/1.2') -> False

register.is_full_name('ds/config/1.0.2') -> True
register.is_full_name('ds/1.0.2') -> True
register.is_full_name('ds_with_number123/1.0.2') -> True
```

Thank you very much.

Reply:
Thanks you so much it will really help so we can get rid of `for fileName in files:`   we can use `tf.iogfile.glob() `for `is_full_name`?

Reply:
`is_full_name` just take a string, not a path, so there is no path / gfile involved. A simple regex should be enough

Reply:
Thank you , I will complete this soon